### PR TITLE
#1: Migrate to themis encryption

### PR DESCRIPTION
**Pull request by vvsevolodovich**

### CODE REVIEW

**Comment by vvsevolodovich** [comment]
PR for #1

**Comment by vixentael** [comment]
Gimme some time to read through, please :)

File: app/src/androidTest/java/com/epam/keystore/ThemisReadWriteInstrumentedTest.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        assertEquals("passWORD", storage.get("key2"));
+        assertEquals("passWORD", storage.get("key1"));
+    }
+
+    @Test
+    public void shouldClearStorage() {
+        storage.save("key12", "1");
+        assertEquals("1", storage.get("key12"));
+        storage.remove("key12");
+        assertNull(storage.get("key12"));
```
Comment (vixentael) [inline app/src/androidTest/java/com/epam/keystore/ThemisReadWriteInstrumentedTest.java]: add another check to make sure that only "key12" is removed
```suggestion
        assertNull(storage.get("key12"));
        
        storage.save("key13", "3456");
        storage.save("key14", "abc");
        storage.remove("key14");
        assertNull(storage.get("key14"));
        assertEquals("3456", storage.get("key13"));
```

File: app/src/androidTest/java/com/epam/keystore/ThemisReadWriteInstrumentedTest.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        storage.remove("KEY");
+        assertNull(storage.get("KEY"));
+    }
+
+    @Test
+    public void shouldClearKeys() {
+        storage.save("KEY", "1");
+        storage.save("KEY2", "2");
+        storage.remove("KEY");
+        assertEquals("2", storage.get("KEY2"));
```
Comment (vixentael) [inline app/src/androidTest/java/com/epam/keystore/ThemisReadWriteInstrumentedTest.java]: should add more keys and make sure that are removed

```suggestion
        assertEquals("2", storage.get("KEY2"));
        assertNull(storage.get("KEY"));
        
        storage.save("KEY3", "3");
        storage.save("KEY4", "4");
```

File: app/src/androidTest/java/com/epam/keystore/ThemisReadWriteInstrumentedTest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        storage.erase();
+        assertNull(storage.get("KEY2"));
```
Comment (vixentael) [inline app/src/androidTest/java/com/epam/keystore/ThemisReadWriteInstrumentedTest.java]: _continue previous_

```suggestion
        assertNull(storage.get("KEY2"));
        assertNull(storage.get("KEY3"));
        assertNull(storage.get("KEY4"));
```

File: app/src/androidTest/java/com/epam/keystore/ThemisReadWriteInstrumentedTest.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    @Test
+    public void shouldSaveValue() {
+        storage.save("key", "passWORD");
+        assertEquals("passWORD", storage.get("key"));
+    }
+
+    @Test
+    public void shouldSaveOtherKeyValue() {
+        storage.save("key1", "passWORD");
+        assertEquals("passWORD", storage.get("key1"));
```
Comment (vixentael) [inline app/src/androidTest/java/com/epam/keystore/ThemisReadWriteInstrumentedTest.java]: what will happens if two values are set to the same key?

like this

```
storage.save("key1", "value1");
storage.save("key1", "value2");
storage.get("key1")
```

should we add test for this?

File: app/src/main/java/com/epam/keystore/providers/themis/ThemisEncryptionProvider.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            return;
+        }
+        if (key != null && value != null) {
+            key = generateKeyWithPrefix(key);
+
+            try {
+                SecureCell sc = new SecureCell(key.getBytes(StandardCharsets.UTF_8), MODE_SEAL);
+                SecureCellData protectedData = sc.protect(key.getBytes(StandardCharsets.UTF_8), value.getBytes(StandardCharsets.UTF_8));
+                String encodedString = Base64.encodeToString(protectedData.getProtectedData(), Base64.NO_WRAP);
+
```
Comment (vixentael) [inline app/src/main/java/com/epam/keystore/providers/themis/ThemisEncryptionProvider.java]: all good here 👍

File: app/src/main/java/com/epam/keystore/providers/themis/ThemisEncryptionProvider.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        try {
+            key = generateKeyWithPrefix(key);
+
+            String encodedString = preferences.getString(key, null);
+            String decryptedData;
+            if (encodedString != null) {
+                byte[] decodedString = Base64.decode(encodedString, Base64.NO_WRAP);
+                SecureCell sc = new SecureCell(key.getBytes(StandardCharsets.UTF_8), MODE_SEAL);
+
+                SecureCellData protectedDataAgain = new SecureCellData(decodedString, null);
```
Comment (vixentael) [inline app/src/main/java/com/epam/keystore/providers/themis/ThemisEncryptionProvider.java]:
```suggestion
                SecureCellData encryptedData = new SecureCellData(decodedString, null);
```

File: app/src/main/java/com/epam/keystore/providers/themis/ThemisEncryptionProvider.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+                byte[] unprotectedData = sc.unprotect(key.getBytes(StandardCharsets.UTF_8), protectedDataAgain);
```
Comment (vixentael) [inline app/src/main/java/com/epam/keystore/providers/themis/ThemisEncryptionProvider.java]:
```suggestion
                byte[] unprotectedData = sc.unprotect(key.getBytes(StandardCharsets.UTF_8), encryptedData);
```

File: themis/build.gradle [hunk 8]
```diff
@@ -0,0 +1,2 @@
+configurations.maybeCreate("default")
+artifacts.add("default", file('themis-release.aar'))
```
Comment (vixentael) [inline themis/build.gradle:2]: if you have any suggestions how to distributed Themis in a more developers-friendly way, please let me know :)

**Review by vixentael — APPROVED** [review]
I put my most attention at tests and `ThemisEncryptionProvider.java`, and I don't have any critical comments. Good job, guys! 👍

From my understanding, you decided to leave CIPHER implementation as is for compatibility with applications that current use this library, yes?

**Review by vixentael — APPROVED** [review]

**Pull request merged**